### ISSUE TITLE
PHPStan integration

### ISSUE THREAD

**Issue by maks-rafalko** [php-decimal/php-decimal#1]
Hello.

When the project uses PHPStan Static Analyzer, we get the following errors because of `php-decimal`.

```php
new Decimal(0) - new Decimal(0);

// Error: Unary operation "-" on Decimal\Decimal results in an error.
```

Does anybody aware of PHPStan plugin that supports `php-decimal` extension and binary operations between objects?

**Comment by rtheunissen** [comment]
This is a great idea. It might affect other extensions too (which overload operators). I wonder how gmp is handled?

**Comment by rtheunissen** [comment]
If there isn't such a plugin, I'll take a look at what is required to write one.

**Comment by maks-rafalko** [comment]
I didn't find it. Please share your thoughts or work here, because I planned to do the same ;) Just to avoid wasting of time

**Comment by rtheunissen** [comment]
I won't be able to make a start until Sunday night PST. But I'll start researching. Let's collaborate here?

**Comment by rtheunissen** [comment]
Also cc @ondrejmirtes

**Comment by maks-rafalko** [comment]
No problem. If I will be able to do it earlier, I will post everything I find here 👍 

Since you mentioned @ondrejmirtes, the question is: what would be a good example of such a plugin to get inspiration from?

**Comment by rtheunissen** [comment]
Exactly, or how could this best be approached? The ideal is to ignore the existing operator rules for certain classes, while being able to infer that the result is an instance of Decimal.

There has been some draft discussions at ext-ds about allowing operator support for sets, ie. a | b being a union, or list concatenation with +. I'm hoping that this will be a good template to implement a similar plugin over there.

**Comment by rtheunissen** [comment]
My starting point is here: https://github.com/phpstan/phpstan/tree/master/src/Rules/Operators

I'll try read into some of the existing plugins though I doubt any of them affect operators.

**Comment by ondrejmirtes** [comment]
Hi, there's no support yet for writing a plugin that would say that arithmetic operations on objects are fine.

Another similar PHP extension like this is bcmath, it's also not supported yet.

I'm not sure what are the semantics of the `Decimal\Decimal` object - can it also be safely cast to `int` or `float`?

**Comment by rtheunissen** [comment]
@ondrejmirtes yes it can be cast to either. Do you mean gmp? As far as I know bc doesn't support operators because it is a functional interface.

What would be required to allow plugins to infer this information? Happy to help.

**Comment by ondrejmirtes** [comment]
It would be best to look into `Scope::resolveType()`, there's logic for `Node\Expr\BinaryOp\Plus`, `Node\Expr\AssignOp\Plus` etc. When it returns `ErrorType`, it's reported as error in these rules.

**Comment by ondrejmirtes** [comment]
I don't think we have to write a plugin, I'm fine with hardcoding this, or using some configuration parameters.

**Comment by rtheunissen** [comment]
Wouldn't that require hard-coding extension-specific behaviour and classes? Would require a core update to stan when behaviour changes. Would it be possible at all to work this into plugin support? Unless that will be significantly more difficult.. I think that's our ideal.

Full disclosure, I haven't used or looked closely at phpstan yet so not sure where the hooks and capabilites lie.

**Comment by rtheunissen** [comment]
The ds list concatenation issue (a + b: list c) is more advanced, so it's not necessarily a numeric result. We should approach this with that in mind.

**Comment by ondrejmirtes** [comment]
I'm willing to also accept a PR if someone comes up with a suitable interface (plugin mechanism) for this :)

**Comment by marcospassos** [comment]
I've opened an issue regarding this thread:
https://github.com/phpstan/phpstan/issues/2000
